### ISSUE TITLE
[Feature] Cross linking packages breaks identity checks in external dependencies

### ISSUE THREAD

**Issue by charlieschwabacher** [lerna/lerna#176]
Imagine I have a lerna repo w/ two packages _a_ and _b_ that depend on each other, and each depend on an external package from npm, _c_.  If I bootstrap and then run code inside _a_, I will end up requiring two different copies of _c_ - the one in `packages/a/node_modules` AND the one in `packages/b/node_modules`.  This breaks any identity checks in _c_.

I'm running into this problem w/ graphql as the external dependency and a failing `instanceof` check.

One way to avoid this problem would be to add shared external dependencies to the root package.json, install them to `node_modules`, and link them from `packages/a/node_modules` and `packages/b/node_modules` in the same way that lerna links the internal dependencies.

Is this a change you would consider?  Or is there another solution for this that I have overlooked?

**Comment by charlieschwabacher** [comment]
I was able to work around my issue by adding a dummy graphql package w/

package.json:
```
{
  "private": true,
  "name": "graphql",
  "version": "0.6.0"
}
```

and index.js:
```
module.exports = require('../../node_modules/graphql');
```

I would still love a cleaner solution, but this is at least a temporary solution.

**Comment by hzoo** [comment]
Will link https://twitter.com/gr2m/status/738454783854563330 as a reference since it seems to be a general problem too

> Is there a CLI tool that lets me git clone multiple repositories, runs npm install and links them based on "dependencies" in package.json?

**Comment by thejameskyle** [comment]
Yeah I've run into this exact issue with React related stuff. I wouldn't mind adding a feature that allowed you to whitelist external dependencies that need to share a single instance that could be linked from the root `node_modules`.

I think we'd have to use actual symlinks though... which we may need to explore again

**Comment by charlieschwabacher** [comment]
@thejameskyle - why do you say we would have to use symlinks?

**Comment by thejameskyle** [comment]
Because you would need to symlink it into each package.

**Comment by rygine** [comment]
this would be fixed using the updated bootstrapping process i've created, as long as they share the same version of the external dependency.

more info here: https://github.com/lerna/lerna/issues/215

**Comment by charlieschwabacher** [comment]
@thejameskyle what I mean is why symlinks here vs proxies for cross linking the local packages?

**Comment by gigabo** [comment]
Oh, this didn't actually wind up getting addressed by #237.

The functionality that was dropped is tracked in #334.

**Comment by ingro** [comment]
I run in this issue as I have multiple packages that depends on `react`. I used @charlieschwabacher solution with a fake react package and it works, but I hope a backed-in solution will come in future!

**Comment by gmmorris** [comment]
Just came across this issue myself after a whole day of debugging a problem >_<
I made a repo for reference and was about to open a new issue when I found this one (See here: https://github.com/gmmorris/potential-symbol-bug).

I took a look at #334 and it's been locked but left open.

Is this something anyone here thinks might be looked at again?
Is it something worth me looking into with a potential pull request?

I'm getting the feeling that if the maintainers have decided not to address on purpose, any pull request would probably get rejected, so just want to double check before I invest time into this.

Thanks for investing your time everyone, I love Lerna and appreciate the work everyone has put into it!

**Comment by gigabo** [comment]
Hi @gmmorris thanks for bringing this back up.  I've opened https://github.com/lerna/lerna/pull/507 to port hoisting back from Asini.

**Comment by gmmorris** [comment]
@gigabo My pleasure, thank *you*. :)
I'll keep an eye on it. I've solved this manually for now, but obviously that's not a good long term solution.

**Comment by SandroMachado** [comment]
Any updates on this issue?

**Comment by photusenigma** [comment]
Also looking for an update.  ?

**Comment by evocateur** [comment]
This was fixed in #507 with the introduction of `--hoist`.
